WordPress/gutenberg
# CustomGradientPicker: Refactor to TypeScript

## What?
Refactor `CustomGradientPicker` component to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744

## Why?

The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

## Testing Instructions
- After applying the patch, run `npm ci`, as a new `DefinitelyTyped` package has been added for the `gradientParser` library this component relies on.
- Ensure all unit and e2e tests still pass.
- In Storybook confirm the component functions as expected and docs tab is loading properly.
- In the editor, test a gradient picker (e.g. on a background color) to ensure it functions as expected.

## Notes:
- There were some interesting challenges with the `DefinitelyTyped` values for the `gradientParser` library. This led to what felt like pretty liberal use of type assertions. Many of these were to avoid conflicts between union members. Anyway, I know there's a bunch of them and I'm very open to feedback them... so anywhere you see me using an `as`, please take it with a grain of salt and don't hesitate to redirect me if you see a better option.
- I've left inline comments where I felt they'd help clarify what I was doing/thinking at the time, but this one felt like it had more moving parts than I'm used to, so let me know if anything seems off.
